internals/search.py

```diff
@@ -259,16 +257,25 @@ def process_query(
   # 3a. Get the result of each future and combine them into a result ID set.
   logging.info('now waiting on futures')
   result_id_set = None
-  for is_or, future in feature_id_future_ops:
+  for logical_op, future in feature_id_future_ops:
     feature_ids = _resolve_promise_to_id_list(future)
     if result_id_set is None:
       logging.info('first term yields %r', feature_ids)
-      result_id_set = set(feature_ids)
-      continue
+      if logical_op == '-':
+        # If a negation is the first term, interpret it as
+        # all feature ids - the negated feature ids. Thus adding
+        # all feature ids here initially.
+        all_feature_entries = core_models.FeatureEntry.query().fetch()
```

Comment:
I think this is correct, but it would be faster if you use `fetch(keys_only=True)` and then get loop over those keys on the next line.

Another small speed up: there is no need to construct the list object before passing it to the set() constructor.   Instead of `set([x for y in z])` it could be `set(x for y in z)`, or even python3 set constructor syntax: `{x for y in z}`.

It could be made even faster by doing it via `fetch_async` and calling it earlier in the process, but that can be left as a TODO.